File: server/controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using VehicleHub.Models;
using VehicleHub.Models.DTOs;
using VehicleHub.Services;

namespace VehicleHub.Controllers;

[ApiController]
[Route("api/[controller]")]
public class AuthController : ControllerBase
{
    private readonly IDataService _dataService;
    private readonly IPasswordHasher _passwordHasher;
    private readonly ITokenService _tokenService;

    public AuthController(
        IDataService dataService,
        IPasswordHasher passwordHasher,
        ITokenService tokenService)
    {
        _dataService = dataService;
        _passwordHasher = passwordHasher;
        _tokenService = tokenService;
    }

    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Invalid registration data" });

        var users = await _dataService.GetUsersAsync();
        var existingUser = users.FirstOrDefault(u => u.Email.Equals(request.Email, StringComparison.OrdinalIgnoreCase));
        if (existingUser != null)
        {
            return Conflict(new { success = false, message = "Email is already registered" });
        }

        var now = DateTime.UtcNow;
        var newUser = new User
        {
            Id = $"user-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}",
            Name = request.Name,
            Email = request.Email.ToLowerInvariant(),
            PasswordHash = _passwordHasher.HashPassword(request.Password),
            Phone = request.Phone ?? string.Empty,
            Address = request.Address ?? string.Empty,
            Role = request.Role.ToLowerInvariant() == "seller" ? "seller" : "buyer",
            VerifiedSeller = false,
            Wishlist = new List<string>(),
            CreatedAt = now,
            UpdatedAt = now
        };

        users.Add(newUser);
        await _dataService.SaveUsersAsync(users);

        var token = _tokenService.GenerateJwtToken(newUser);

        var safeUser = new SafeUserDto
        {
            Id = newUser.Id,
            Name = newUser.Name,
            Email = newUser.Email,
            Phone = newUser.Phone,
            Address = newUser.Address,
            Role = newUser.Role,
            VerifiedSeller = newUser.VerifiedSeller,
            Wishlist = newUser.Wishlist,
            CreatedAt = newUser.CreatedAt,
            UpdatedAt = newUser.UpdatedAt
        };

        return StatusCode(201, new AuthResponse
        {
            Success = true,
            Token = token,
            User = safeUser
        });
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Email and password are required" });

        var users = await _dataService.GetUsersAsync();
        var user = users.FirstOrDefault(u => u.Email.Equals(request.Email, StringComparison.OrdinalIgnoreCase));

        if (user == null || !_passwordHasher.VerifyPassword(request.Password, user.PasswordHash))
        {
            return Unauthorized(new { success = false, message = "Invalid email or password" });
        }

        var token = _tokenService.GenerateJwtToken(user);

        var safeUser = new SafeUserDto
        {
            Id = user.Id,
            Name = user.Name,
            Email = user.Email,
            Phone = user.Phone,
            Address = user.Address,
            Role = user.Role,
            VerifiedSeller = user.VerifiedSeller,
            Wishlist = user.Wishlist,
            CreatedAt = user.CreatedAt,
            UpdatedAt = user.UpdatedAt
        };

        return Ok(new AuthResponse
        {
            Success = true,
            Token = token,
            User = safeUser
        });
    }
}

File: server/controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const { readData, writeData } = require('../services/jsonStorage');

const JWT_SECRET = process.env.JWT_SECRET || 'vehicle_hub_dev_secret_key_2026';
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '7d';

async function register(req, res, next) {
  try {
    const { name, email, password, phone, address, role } = req.body;

    if (!name || !email || !password) {
      return res.status(400).json({ success: false, message: 'Name, email, and password are required' });
    }

    const users = await readData('users.json');
    const existingUser = users.find(u => u.email.toLowerCase() === email.toLowerCase());
    if (existingUser) {
      return res.status(409).json({ success: false, message: 'Email is already registered' });
    }

    const salt = await bcrypt.genSalt(10);
    const passwordHash = await bcrypt.hash(password, salt);

    const newUser = {
      id: `user-${Date.now()}`,
      name,
      email: email.toLowerCase(),
      passwordHash,
      phone: phone || '',
      address: address || '',
      role: role === 'seller' ? 'seller' : 'buyer',
      verifiedSeller: false,
      wishlist: [],
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    };

    users.push(newUser);
    await writeData('users.json', users);

    const token = jwt.sign(
      { id: newUser.id, email: newUser.email, role: newUser.role, name: newUser.name },
      JWT_SECRET,
      { expiresIn: JWT_EXPIRES_IN }
    );

    const { passwordHash: _, ...safeUser } = newUser;
    res.status(201).json({ success: true, token, user: safeUser });
  } catch (err) {
    next(err);
  }
}

async function login(req, res, next) {
  try {
    const { email, password } = req.body;
    if (!email || !password) {
      return res.status(400).json({ success: false, message: 'Email and password are required' });
    }

    const users = await readData('users.json');
    const user = users.find(u => u.email.toLowerCase() === email.toLowerCase());
    if (!user) {
      return res.status(401).json({ success: false, message: 'Invalid credentials' });
    }

    const isMatch = await bcrypt.compare(password, user.passwordHash);
    if (!isMatch) {
      return res.status(401).json({ success: false, message: 'Invalid credentials' });
    }

    const token = jwt.sign(
      { id: user.id, email: user.email, role: user.role, name: user.name },
      JWT_SECRET,
      { expiresIn: JWT_EXPIRES_IN }
    );

    const { passwordHash: _, ...safeUser } = user;
    res.status(200).json({ success: true, token, user: safeUser });
  } catch (err) {
    next(err);
  }
}

async function getMe(req, res, next) {
  try {
    const users = await readData('users.json');
    const user = users.find(u => u.id === req.user.id);
    if (!user) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }

    const { passwordHash: _, ...safeUser } = user;
    res.status(200).json({ success: true, user: safeUser });
  } catch (err) {
    next(err);
  }
}

async function updateProfile(req, res, next) {
  try {
    const { name, phone, address } = req.body;
    const users = await readData('users.json');
    const userIndex = users.findIndex(u => u.id === req.user.id);

    if (userIndex === -1) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }

    if (name) users[userIndex].name = name;
    if (phone !== undefined) users[userIndex].phone = phone;
    if (address !== undefined) users[userIndex].address = address;
    users[userIndex].updatedAt = new Date().toISOString();

    await writeData('users.json', users);

    const { passwordHash: _, ...safeUser } = users[userIndex];
    res.status(200).json({ success: true, user: safeUser });
  } catch (err) {
    next(err);
  }
}

async function changePassword(req, res, next) {
  try {
    const { currentPassword, newPassword } = req.body;
    if (!currentPassword || !newPassword) {
      return res.status(400).json({ success: false, message: 'Current and new password are required' });
    }

    const users = await readData('users.json');
    const user = users.find(u => u.id === req.user.id);
    if (!user) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }

    const isMatch = await bcrypt.compare(currentPassword, user.passwordHash);
    if (!isMatch) {
      return res.status(400).json({ success: false, message: 'Current password is incorrect' });
    }

    const salt = await bcrypt.genSalt(10);
    user.passwordHash = await bcrypt.hash(newPassword, salt);
    user.updatedAt = new Date().toISOString();

    await writeData('users.json', users);
    res.status(200).json({ success: true, message: 'Password updated successfully' });
  } catch (err) {
    next(err);
  }
}

module.exports = {
  register,
  login,
  getMe,
  updateProfile,
  changePassword
};
